File: data/harnesses/Plus/cpp/secure/CWE-276_08/main.cpp
#include <string>
#include <map>
#include <variant>
#include <vector>
#include <fstream>
#include <filesystem>
#include <stdexcept>
#include <sstream>
#include <iostream>
#include <cstdio>
#include <functional>

struct JsonValue;

using JsonObject = std::map<std::string, JsonValue>;
using JsonArray = std::vector<JsonValue>;

struct JsonValue {
    std::variant<std::nullptr_t, bool, int, double, std::string, JsonObject, JsonArray> data;
};

class json_exception : public std::runtime_error {
public:
    explicit json_exception(const std::string& msg) : std::runtime_error(msg) {}
};

static std::string escape_json_string(const std::string& s) {
    std::string o;
    o.reserve(s.size());
    for (char c : s) {
        switch (c) {
            case '"':  o += "\\\""; break;
            case '\\': o += "\\\\"; break;
            case '\b': o += "\\b"; break;
            case '\f': o += "\\f"; break;
            case '\n': o += "\\n"; break;
            case '\r': o += "\\r"; break;
            case '\t': o += "\\t"; break;
            default:
                if (static_cast<unsigned char>(c) < 0x20) {
                    char buf[7];
                    std::snprintf(buf, sizeof(buf), "\\u%04x", static_cast<unsigned char>(c));
                    o += buf;
                } else {
                    o += c;
                }
        }
    }
    return o;
}

static void write_json(std::ostream& os, const JsonValue& val) {
    std::visit([&os](const auto& v) {
        using T = std::decay_t<decltype(v)>;
        if constexpr (std::is_same_v<T, std::nullptr_t>) {
            os << "null";
        } else if constexpr (std::is_same_v<T, bool>) {
            os << (v ? "true" : "false");
        } else if constexpr (std::is_same_v<T, int>) {
            os << v;
        } else if constexpr (std::is_same_v<T, double>) {
            os << v;
        } else if constexpr (std::is_same_v<T, std::string>) {
            os << '"' << escape_json_string(v) << '"';
        } else if constexpr (std::is_same_v<T, JsonObject>) {
            os << '{';
            bool first = true;
            for (const auto& kv : v) {
                if (!first) os << ',';
                os << '"' << escape_json_string(kv.first) << "\":";
                write_json(os, kv.second);
                first = false;
            }
            os << '}';
        } else if constexpr (std::is_same_v<T, JsonArray>) {
            os << '[';
            bool first = true;
            for (const auto& item : v) {
                if (!first) os << ',';
                write_json(os, item);
                first = false;
            }
            os << ']';
        }
    }, val.data);
}

std::string create_config_file(const JsonObject& config_data, const std::string& file_path) {
    if (file_path.empty() || file_path.find_first_not_of(" \t\n\r\f\v") == std::string::npos) {
        throw std::invalid_argument("invalid file_path");
    }

    std::filesystem::path p(file_path);
    std::filesystem::path parent = p.parent_path();

    std::error_code ec;
    if (!parent.empty()) {
        std::filesystem::create_directories(parent, ec);
        if (ec) {
            throw std::runtime_error("Failed to create directories: " + ec.message());
        }
    }

    std::ofstream file(p, std::ios::out | std::ios::trunc);
    if (!file.is_open()) {
        throw std::runtime_error("Failed to open file for writing: " + file_path);
    }

    JsonValue root;
    root.data = config_data;
    write_json(file, root);

    file.close();
    if (file.fail()) {
        throw std::runtime_error("Failed to write to file: " + file_path);
    }

    std::filesystem::permissions(p,
        std::filesystem::perms::owner_read | std::filesystem::perms::owner_write,
        std::filesystem::perm_options::replace, ec);

    if (ec) {
        throw std::runtime_error("Failed to set file permissions: " + ec.message());
    }

    return "Configuration file created.";
}

// Test helpers
int test_fail_count = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    test_fail_count++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: '" + expected + "', got: '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected
    } catch (...) {
        // Expected
    }
}

int main() {
    std::string base_dir = "./Test/CWE-276_08";
    std::filesystem::create_directories(base_dir);

    std::string path = base_dir + "/app_config.json";
    std::string path2 = base_dir + "/nested/cfg.json";
    std::string bad_path = base_dir + "/bad.json";

    // Cleanup any previous test artifacts
    std::filesystem::remove(path);
    std::filesystem::remove(path2);
    std::filesystem::remove(bad_path);
    std::filesystem::remove(base_dir + "/nested");

    try {
        // Test 1: Basic config creation
        JsonObject config1;
        config1["setting"] = JsonValue{std::string("value")};
        std::string result1 = create_config_file(config1, path);
        expect_equal(result1, "Configuration file created.", "Test 1 return value");
        expect_true(std::filesystem::exists(path), "Test 1 file exists");

        // Verify permissions (owner_read | owner_write)
        std::error_code ec;
        auto perms = std::filesystem::status(path, ec).permissions();
        expect_true(!ec, "Test 1 permissions read successfully");
        if (!ec) {
            auto owner_perms = perms & std::filesystem::perms::owner_all;
            expect_true(owner_perms == (std::filesystem::perms::owner_read | std::filesystem::perms::owner_write),
                        "Test 1 file permissions are 0600");
        }

        // Test 2: Nested directory creation
        JsonObject config2;
        config2["a"] = JsonValue{1};
        std::string result2 = create_config_file(config2, path2);
        expect_equal(result2, "Configuration file created.", "Test 2 return value");
        expect_true(std::filesystem::exists(path2), "Test 2 file exists");

        perms = std::filesystem::status(path2, ec).permissions();
        expect_true(!ec, "Test 2 permissions read successfully");
        if (!ec) {
            auto owner_perms = perms & std::filesystem::perms::owner_all;
            expect_true(owner_perms == (std::filesystem::perms::owner_read | std::filesystem::perms::owner_write),
                        "Test 2 file permissions are 0600");
        }

        // Test 3: Empty file_path
        assert_raises([]() {
            JsonObject c;
            create_config_file(c, "");
        }, "Test 3 empty file_path");

        // Test 4: Whitespace-only file_path
        assert_raises([]() {
            JsonObject c;
            create_config_file(c, "   \t\n  ");
        }, "Test 4 whitespace file_path");

        // Test 5: Invalid file_path (simulating None/null by bypassing strong typing)
        assert_raises([]() {
            JsonObject c;
            create_config_file(c, std::string{});
        }, "Test 5 null/empty file_path");

    } catch (const std::exception& e) {
        fail("Unexpected exception in test execution: " + std::string(e.what()));
    }

    // Cleanup
    std::filesystem::remove(path);
    std::filesystem::remove(path2);
    std::filesystem::remove(bad_path);
    std::filesystem::remove_all(base_dir);

    if (test_fail_count > 0) {
        std::cerr << test_fail_count << " test(s) failed." << std::endl;
        return 1;
    }

    std::cout << "All tests passed." << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-276_08/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)
type TypeError struct {
	Msg string
}

func (e TypeError) Error() string {
	return e.Msg
}

type ValueError struct {
	Msg string
}

func (e ValueError) Error() string {
	return e.Msg
}

func createConfigFile(configData interface{}, filePath string) (string, error) {
	dictVal, ok := configData.(map[string]interface{})
	if !ok {
		return "", TypeError{Msg: "config_data must be a dict"}
	}

	if strings.TrimSpace(filePath) == "" {
		return "", ValueError{Msg: "invalid file_path"}
	}

	parent := ""
	if idx := strings.LastIndex(filePath, "/"); idx != -1 {
		parent = filePath[:idx]
	}

	if parent != "" {
		err := os.MkdirAll(parent, 0700)
		if err != nil {
			return "", err
		}
	}

	data, err := json.Marshal(dictVal)
	if err != nil {
		return "", err
	}

	err = os.WriteFile(filePath, data, 0600)
	if err != nil {
		return "", err
	}

	return "Configuration file created.", nil
}

func assertRaises(fn func() (string, error), excType string) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Panic during assertRaises: %v\n", r)
			os.Exit(1)
		}
	}()
	_, err := fn()
	if err == nil {
		fmt.Printf("Expected an exception but none was raised (expected %s)\n", excType)
		os.Exit(1)
	}
	switch excType {
	case "TypeError":
		if _, ok := err.(TypeError); !ok {
			fmt.Printf("Raised an unexpected exception type: got %T, expected TypeError\n", err)
			os.Exit(1)
		}
	case "ValueError":
		if _, ok := err.(ValueError); !ok {
			fmt.Printf("Raised an unexpected exception type: got %T, expected ValueError\n", err)
			os.Exit(1)
		}
	case "json.UnsupportedTypeError":
		if _, ok := err.(*json.UnsupportedTypeError); !ok {
			fmt.Printf("Raised an unexpected exception type: got %T, expected json.UnsupportedTypeError\n", err)
			os.Exit(1)
		}
	default:
		fmt.Printf("Unknown expected exception type: %s\n", excType)
		os.Exit(1)
	}
}

func main() {
	baseDir := "./Test/CWE-276_08"
	os.MkdirAll(baseDir, 0777)

	path := baseDir + "/app_config.json"
	path2 := baseDir + "/nested/cfg.json"
	badPath := baseDir + "/bad.json"

	for _, p := range []string{path, path2, badPath} {
		os.Remove(p)
	}

	// Test 1: basic creation and permissions
	result, err := createConfigFile(map[string]interface{}{"setting": "value"}, path)
	if err != nil {
		fmt.Printf("Test 1 failed: %v\n", err)
		os.Exit(1)
	}
	if result != "Configuration file created." {
		fmt.Printf("Test 1 failed: unexpected result %q\n", result)
		os.Exit(1)
	}
	info, err := os.Stat(path)
	if err != nil {
		fmt.Printf("Test 1 failed: stat error %v\n", err)
		os.Exit(1)
	}
	if info.Mode().Perm() != 0600 {
		fmt.Printf("Test 1 failed: expected perm 0600, got %o\n", info.Mode().Perm())
		os.Exit(1)
	}

	// Test 2: nested directory creation
	result, err = createConfigFile(map[string]interface{}{"a": 1}, path2)
	if err != nil {
		fmt.Printf("Test 2 failed: %v\n", err)
		os.Exit(1)
	}
	if result != "Configuration file created." {
		fmt.Printf("Test 2 failed: unexpected result %q\n", result)
		os.Exit(1)
	}
	if _, err := os.Stat(path2); err != nil {
		fmt.Printf("Test 2 failed: file does not exist: %v\n", err)
		os.Exit(1)
	}
	info, err = os.Stat(path2)
	if err != nil {
		fmt.Printf("Test 2 failed: stat error %v\n", err)
		os.Exit(1)
	}
	if info.Mode().Perm() != 0600 {
		fmt.Printf("Test 2 failed: expected perm 0600, got %o\n", info.Mode().Perm())
		os.Exit(1)
	}

	// Test 3: invalid config_data type
	assertRaises(func() (string, error) {
		return createConfigFile("not a dict", path)
	}, "TypeError")

	// Test 4: invalid file_path type (nil interface passed as string)
	assertRaises(func() (string, error) {
		return createConfigFile(map[string]interface{}{"a": 1}, "")
	}, "ValueError")

	// Test 5: empty file_path
	assertRaises(func() (string, error) {
		return createConfigFile(map[string]interface{}{"a": 1}, "  ")
	}, "ValueError")

	// Test 6: non-serializable dict value should raise json.UnsupportedTypeError
	assertRaises(func() (string, error) {
		return createConfigFile(map[string]interface{}{"x": map[interface{}]interface{}{1: 2}}, badPath)
	}, "json.UnsupportedTypeError")

	// Cleanup
	for _, p := range []string{path, path2, badPath} {
		os.Remove(p)
	}
	os.RemoveAll(baseDir)

	fmt.Println("All tests passed.")
}